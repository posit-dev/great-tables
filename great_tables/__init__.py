# Set version ----
from importlib_metadata import version as _v

__version__ = _v("great_tables")

del _v

# Main gt imports ----

from .gt import GT
from . import vals, loc, style
from ._styles import FromColumn as from_column
from ._helpers import (
    letters,
    LETTERS,
    px,
    pct,
    md,
    html,
    google_font,
    random_id,
    system_fonts,
    define_units,
    nanoplot_options,
)


__all__ = (
    "GT",
    "exibble",
    "letters",
    "LETTERS",
    "px",
    "pct",
    "md",
    "html",
    "google_font",
    "system_fonts",
    "define_units",
    "nanoplot_options",
    "random_id",
    "from_column",
    "vals",
    "loc",
    "style",
)


def __getattr__(k: str):
    # exibble dataset available on top-level module, but is a pandas DataFrame.
    # Since pandas is an optional dependency, we import exibble dynamically.
    if k == "exibble":
        from great_tables.data import exibble

        return exibble

    # allow the data submodule to be accessed directly, as if it were a top-level import
    if k == "data":
        import great_tables.data

        return great_tables.data

    raise AttributeError(f"cannot get attribute {k} from great_tables ({__file__})")
