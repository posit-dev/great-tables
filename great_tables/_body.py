from __future__ import annotations

from typing import TYPE_CHECKING

from ._tbl_data import copy_data

if TYPE_CHECKING:
    from ._gt_data import Body


def body_reassemble(body: Body) -> Body:
    # Note that this used to order the body based on groupings, but now that occurs in the
    # renderer itself.
    return body.__class__(copy_data(body.body))
