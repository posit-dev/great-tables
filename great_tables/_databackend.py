from __future__ import annotations

import importlib
from abc import ABCMeta
from types import ModuleType


def _load_class(mod_name: str, cls_name: str) -> type:
    # follow the path of mods if it is one
    listified_mod_path: list[str] = cls_name.split(".")

    sub_module_path: list[str] = [mod_name] + listified_mod_path[:-1]

    cur_mod: ModuleType = importlib.import_module(sub_module_path[0])
    _cls_name: str = listified_mod_path[-1]

    if len(sub_module_path) == 1:  # no submodules, just the module itself
        return getattr(cur_mod, _cls_name)

    # now, we must increment the module paths to import the final class
    remaining_mods: list[str] = sub_module_path[1:]
    for mod in remaining_mods:
        cur_mod: ModuleType = getattr(cur_mod, mod)

    return getattr(cur_mod, _cls_name)


class _AbstractBackendMeta(ABCMeta):
    def register_backend(cls, mod_name: str, cls_name: str):
        cls._backends.append((mod_name, cls_name))
        cls._abc_caches_clear()


class AbstractBackend(metaclass=_AbstractBackendMeta):
    @classmethod
    def __init_subclass__(cls):
        if not hasattr(cls, "_backends"):
            cls._backends = []

    @classmethod
    def __subclasshook__(cls, subclass):
        for mod_name, cls_name in cls._backends:
            try:
                parent_candidate = _load_class(mod_name, cls_name)
                if issubclass(subclass, parent_candidate):
                    return True
            except (ModuleNotFoundError, ImportError):
                continue

        return NotImplemented
