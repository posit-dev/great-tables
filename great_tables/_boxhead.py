from __future__ import annotations

from typing import TYPE_CHECKING

from ._locations import resolve_cols_c
from ._utils import _assert_list_is_subset
from ._tbl_data import SelectExpr
from ._text import BaseText

if TYPE_CHECKING:
    from ._types import GTSelf


def cols_label(
    self: GTSelf, cases: dict[str, str | BaseText] | None = None, **kwargs: str | BaseText
) -> GTSelf:
    """
    Relabel one or more columns.

    There are three important pieces to labelling:

    * Each argument has the form: {name in data} = {new label}.
    * Multiple columns may be given the same label.
    * Labels may use curly braces to apply special formatting, called unit notation.
      For example, "area ({{ft^2}})" would appear as "area (ftÂ²)".

    See [`define_units()`](`great_tables.define_units`) for details on unit notation.

    Parameters
    ----------
    cases
        A dictionary where the keys are column names and the values are the labels. Labels may use
        [`md()`](`great_tables.md`) or [`html()`](`great_tables.html`) helpers for formatting.

    **kwargs
        Keyword arguments to specify column labels. Each keyword corresponds to a column name, with
        its value indicating the new label.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Notes
    -----
    GT always selects columns using their name in the underlying data. This means that a column's
    label is purely for final presentation.

    Examples
    --------

    The example below relabels columns from the `countrypops` data to start with uppercase.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "Uganda"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Country Name",
            year="Year",
            population="Population"
        )
    )
    ```

    Note that we supplied the name of the column as the key, and the new label as the value.

    We can also use Markdown formatting for the column labels. In this example, we'll use
    `md("*Population*")` to make the label italicized.

    ```{python}
    from great_tables import GT, md
    from great_tables.data import countrypops

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Name",
            year="Year",
            population=md("*Population*")
        )
    )
    ```

    We can also use unit notation to format the column labels. In this example, we'll use
    `{{cm^3 molecules^-1 s^-1}}` for part of the label for the `OH_k298` column.

    ```{python}
    from great_tables import GT
    from great_tables.data import reactions
    import polars as pl

    reactions_mini = (
        pl.from_pandas(reactions)
        .filter(pl.col("cmpd_type") == "mercaptan")
        .select(["cmpd_name", "OH_k298"])
    )

    (
        GT(reactions_mini)
        .fmt_scientific("OH_k298")
        .sub_missing()
        .cols_label(
            cmpd_name="Compound Name",
            OH_k298="OH, {{cm^3 molecules^-1 s^-1}}",
        )
    )
    ```
    """
    from great_tables._helpers import UnitStr

    cases = cases if cases is not None else {}
    new_cases = cases | kwargs

    # If nothing is provided, return `data` unchanged
    if len(new_cases) == 0:
        return self

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()
    mod_columns = list(new_cases.keys())

    # Stop function if any of the column names specified are not in `cols_labels`
    # msg: "All column names provided must exist in the input `.data` table."
    _assert_list_is_subset(mod_columns, set_list=column_names)

    # Handle units syntax in labels (e.g., "Density ({{ppl / mi^2}})")
    new_kwargs: dict[str, UnitStr | str | BaseText] = {}

    for k, v in new_cases.items():
        if isinstance(v, str):
            unitstr_v = UnitStr.from_str(v)

            if len(unitstr_v.units_str) == 1 and isinstance(unitstr_v.units_str[0], str):
                new_kwargs[k] = unitstr_v.units_str[0]
            else:
                new_kwargs[k] = unitstr_v

        elif isinstance(v, BaseText):
            new_kwargs[k] = v

        else:
            raise ValueError(
                "Column labels must be strings or BaseText objects. Use `md()` or `html()` for formatting."
            )

    boxhead = self._boxhead._set_column_labels(new_kwargs)

    return self._replace(_boxhead=boxhead)


def cols_align(self: GTSelf, align: str = "left", columns: SelectExpr = None) -> GTSelf:
    """
    Set the alignment of one or more columns.

    The `cols_align()` method sets the alignment of one or more columns. The `align` argument
    can be set to one of `"left"`, `"center"`, or `"right"` and the `columns` argument can be
    used to specify which columns to apply the alignment to. If `columns` is not specified, the
    alignment is applied to all columns.

    Parameters
    ----------
    align
        The alignment to apply. Must be one of `"left"`, `"center"`, or `"right"`.
    columns
        The columns to target. Can either be a single column name or a series of column names
        provided in a list. If `None`, the alignment is applied to all columns.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Examples
    --------
    Let's use the `countrypops` to create a small table. We can change the alignment of the
    `population` column with `cols_align()`. In this example, the column label and body cells of
    `population` will be aligned to the left.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "San Marino"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini, rowname_col="year", groupname_col="country_name")
        .cols_align(align="left", columns="population")
    )
    ```

    """

    # Throw if `align` is not one of the three allowed values
    if align not in ["left", "center", "right"]:
        raise ValueError("Align must be one of 'left', 'center', or 'right'.")

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()

    # Upgrade `columns` to a list if `columns` is a string and not None
    if isinstance(columns, str):
        columns = [columns]
        _assert_list_is_subset(columns, set_list=column_names)
    elif columns is None:
        columns = column_names

    sel_cols = resolve_cols_c(data=self, expr=columns)

    # Set the alignment for each column
    return self._replace(_boxhead=self._boxhead._set_column_aligns(sel_cols, align=align))
