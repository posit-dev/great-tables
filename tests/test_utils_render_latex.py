import pytest
from unittest import mock
import pandas as pd
import os

from great_tables import GT, exibble
from great_tables.data import gtcars

from great_tables._utils_render_latex import (
    is_css_length_string,
    is_number_without_units,
    css_length_has_supported_units,
    get_px_conversion,
    get_units_from_length_string,
    convert_to_px,
    convert_to_pt,
    create_wrap_start_l,
    create_fontsize_statement_l,
    create_heading_component_l,
    create_body_component_l,
    create_columns_component_l,
    create_footer_component_l,
    create_wrap_end_l,
    create_table_end_l,
    create_table_start_l,
    derive_table_width_statement_l,
    _render_as_latex,
)


@pytest.fixture
def gt_tbl():
    return GT(pd.DataFrame({"x": [1, 2], "y": [4, 5]}))


@pytest.fixture
def gt_tbl_dec():
    return GT(pd.DataFrame({"x": [1.52, 2.23], "y": [4.75, 5.23]}))


@pytest.fixture
def gt_tbl_sci():
    return GT(pd.DataFrame({"x": [465633.46, -0.00000000345], "y": [4.509, 176.23]}))


@pytest.fixture
def gt_tbl_pct():
    return GT(pd.DataFrame({"x": [0.53, 0.0674], "y": [0.17, 0.32]}))


@pytest.fixture
def gt_tbl_dttm():
    return GT(
        pd.DataFrame(
            {
                "date": ["2023-08-12", "2020-11-17"],
                "time": ["09:21:23", "22:45:02"],
                "dttm": ["2023-08-12 09:21:23", "2020-11-17 22:45:02"],
            }
        )
    )


def test_is_css_length_string():
    assert is_css_length_string("12.5pt")
    assert is_css_length_string("12.5px")
    assert is_css_length_string("12.5")
    assert is_css_length_string("12.5units")


def test_is_number_without_units():
    assert is_number_without_units("12.5")
    assert not is_number_without_units("12.5pt")


def test_css_length_has_supported_units():
    assert css_length_has_supported_units("12.5pt")
    assert css_length_has_supported_units("12.5px")
    assert css_length_has_supported_units("12.5")
    assert css_length_has_supported_units("12.5pt", no_units_valid=False)
    assert css_length_has_supported_units("12.5px", no_units_valid=False)
    assert not css_length_has_supported_units("12.5", no_units_valid=False)
    assert not css_length_has_supported_units("12.8units")
    assert not css_length_has_supported_units("units12.8")


def test_get_units_from_length_string():
    assert get_units_from_length_string("12.5pt") == "pt"
    assert get_units_from_length_string("") == "px"


def test_get_px_conversion_val():
    assert get_px_conversion(length="2343.23pt") == 4 / 3
    assert get_px_conversion(length="43.2px") == 1.0


def test_get_px_conversion_val_raises():
    with pytest.raises(ValueError) as exc_info:
        get_px_conversion(length="12.8bolts")

    assert "Invalid units: bolts" in exc_info.value.args[0]


def test_convert_to_px():
    assert convert_to_px("12.5pt") == 17.0
    assert convert_to_px("12.5px") == 12.5


def test_convert_to_pt():
    assert convert_to_pt("16px") == 12.0


def test_create_fontsize_statement_l(gt_tbl: GT):
    assert create_fontsize_statement_l(gt_tbl) == "\\fontsize{12.0pt}{14.4pt}\\selectfont\n"


def test_create_fontsize_statement_l_pt(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_font_size="18.2pt")

    assert create_fontsize_statement_l(gt_tbl_new) == "\\fontsize{18.2pt}{21.8pt}\\selectfont\n"


def test_create_fontsize_statement_l_px(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_font_size="11px")

    assert create_fontsize_statement_l(gt_tbl_new) == "\\fontsize{8.2pt}{9.9pt}\\selectfont\n"


def test_create_fontsize_statement_l_pct(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_font_size="50%")

    assert create_fontsize_statement_l(gt_tbl_new) == "\\fontsize{6.0pt}{7.2pt}\\selectfont\n"


def test_create_fontsize_statement_l_cm(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_font_size="0.6cm")

    assert create_fontsize_statement_l(gt_tbl_new) == "\\fontsize{17.2pt}{20.7pt}\\selectfont\n"


def test_create_fontsize_statement_l_unknown_unit(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_font_size="1span")

    assert create_fontsize_statement_l(gt_tbl_new) == ""


def test_derive_table_width_statement_l_px_lt(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_width="500px")

    assert (
        derive_table_width_statement_l(gt_tbl_new, use_longtable=True)
        == "\\setlength\\LTleft{\\dimexpr(0.5\\linewidth - 187.5pt)}\n\\setlength\\LTright{\\dimexpr(0.5\\linewidth - 187.5pt)}"
    )


def test_derive_table_width_statement_l_pct_lt(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_width="45%")

    assert (
        derive_table_width_statement_l(gt_tbl_new, use_longtable=True)
        == "\\setlength\\LTleft{0.275\\linewidth}\n\\setlength\\LTright{0.275\\linewidth}"
    )


def test_derive_table_width_statement_l_px_no_lt(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_width="500px")

    assert derive_table_width_statement_l(gt_tbl_new, use_longtable=False) == ""


def test_create_fontsize_statement_l_settings():
    gt_tbl = GT(exibble).tab_options(table_font_size="18.5px")

    assert create_fontsize_statement_l(gt_tbl) == "\\fontsize{13.9pt}{16.6pt}\\selectfont\n"


def test_create_heading_component_l():
    gt_tbl_no_heading = GT(exibble)
    gt_tbl_title = GT(exibble).tab_header(title="Title")
    gt_tbl_title_subtitle = GT(exibble).tab_header(title="Title", subtitle="Subtitle")

    assert create_heading_component_l(gt_tbl_no_heading, use_longtable=False) == ""
    assert (
        create_heading_component_l(gt_tbl_title, use_longtable=False)
        == "\\caption*{\n{\\large Title}\n} "
    )
    assert (
        create_heading_component_l(gt_tbl_title_subtitle, use_longtable=False)
        == "\\caption*{\n{\\large Title} \\\\\n{\\small Subtitle}\n} "
    )


def test_create_columns_component_l_simple():
    gt_tbl = GT(exibble)

    assert (
        create_columns_component_l(data=gt_tbl)
        == "\\toprule\nnum & char & fctr & date & time & datetime & currency & row & group \\\\ \n\\midrule\\addlinespace[2.5pt]"
    )


def test_create_columns_component_l_simple_hidden_cols():
    gt_tbl = GT(exibble).cols_hide(columns=["char", "date"])

    assert (
        create_columns_component_l(data=gt_tbl)
        == "\\toprule\nnum & fctr & time & datetime & currency & row & group \\\\ \n\\midrule\\addlinespace[2.5pt]"
    )


def test_create_columns_component_l_one_spanner():
    gt_tbl = GT(exibble).tab_spanner(label="Spanner", columns=["num", "char"])

    assert (
        create_columns_component_l(data=gt_tbl)
        == "\\toprule\n\\multicolumn{2}{c}{Spanner} &  \\\\ \n\\cmidrule(lr){1-2}\nnum & char & fctr & date & time & datetime & currency & row & group \\\\ \n\\midrule\\addlinespace[2.5pt]"
    )


def test_create_columns_component_l_adjacent_spanners_hiding():
    gt_tbl = (
        GT(exibble)
        .tab_spanner(label="Spanner 1", columns=["num", "char"])
        .tab_spanner(label="Spanner 2", columns=["date", "time"])
        .tab_spanner(label="Spanner 3", columns=["currency", "row"])
        .cols_hide(columns="row")
    )

    assert (
        create_columns_component_l(data=gt_tbl)
        == "\\toprule\n\\multicolumn{2}{c}{Spanner 1} &  & \\multicolumn{2}{c}{Spanner 2} &  & \\multicolumn{1}{c}{Spanner 3} &  \\\\ \n\\cmidrule(lr){1-2} \\cmidrule(lr){4-5} \\cmidrule(lr){7-7}\nnum & char & fctr & date & time & datetime & currency & group \\\\ \n\\midrule\\addlinespace[2.5pt]"
    )


def test_create_columns_component_l_many_spanners():
    gt_tbl = (
        GT(exibble)
        .tab_spanner(label="Spanner 1", columns=["num", "char"])
        .tab_spanner(label="Spanner 2", columns=["date", "time"])
        .tab_spanner(label="Spanner 3", columns=["currency", "row"])
        .tab_spanner(label="Spanner Above 1", columns=["char", "fctr"])
        .tab_spanner(label="Spanner Above 2", columns=["time", "datetime"])
    )

    assert (
        create_columns_component_l(data=gt_tbl)
        == "\\toprule\n & \\multicolumn{2}{c}{Spanner Above 1} &  & \\multicolumn{2}{c}{Spanner Above 2} &  \\\\ \n\\cmidrule(lr){2-3} \\cmidrule(lr){5-6}\n\\multicolumn{2}{c}{Spanner 1} &  & \\multicolumn{2}{c}{Spanner 2} &  & \\multicolumn{2}{c}{Spanner 3} &  \\\\ \n\\cmidrule(lr){1-2} \\cmidrule(lr){4-5} \\cmidrule(lr){7-8}\nnum & char & fctr & date & time & datetime & currency & row & group \\\\ \n\\midrule\\addlinespace[2.5pt]"
    )


def test_create_body_component_l_simple(gt_tbl: GT):
    assert create_body_component_l(data=gt_tbl) == "1 & 4 \\\\\n2 & 5 \\\\"


def test_create_footer_component_one_note(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_source_note(source_note="Source Note.")

    assert (
        create_footer_component_l(gt_tbl_new)
        == "\\begin{minipage}{\\linewidth}\nSource Note.\\\\\n\\end{minipage}"
    )


def test_create_footer_component_two_notes(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_source_note(source_note="Source Note 1.").tab_source_note(
        source_note="Source Note 2."
    )

    assert (
        create_footer_component_l(gt_tbl_new)
        == "\\begin{minipage}{\\linewidth}\nSource Note 1.\\\\\nSource Note 2.\\\\\n\\end{minipage}"
    )


def test_create_footer_component_no_notes(gt_tbl: GT):
    assert create_footer_component_l(gt_tbl) == ""


def test_create_body_component_l_fmt_number(gt_tbl_dec: GT):
    gt_tbl_built = gt_tbl_dec.fmt_number(
        columns="x", rows=0, scale_by=-1, decimals=3, pattern="{x} _"
    )._build_data(context="latex")

    assert create_body_component_l(data=gt_tbl_built) == "-1.520 \\_ & 4.75 \\\\\n2.23 & 5.23 \\\\"


def test_create_body_component_l_fmt_integer(gt_tbl_dec: GT):
    gt_tbl_built = gt_tbl_dec.fmt_integer(
        columns="x", rows=0, scale_by=-1, pattern="{x} _"
    )._build_data(context="latex")

    assert create_body_component_l(data=gt_tbl_built) == "-2 \\_ & 4.75 \\\\\n2.23 & 5.23 \\\\"


def test_create_body_component_l_fmt_scientific(gt_tbl_sci: GT):
    gt_tbl_built = gt_tbl_sci.fmt_scientific(columns="x", pattern="{x} _")._build_data(
        context="latex"
    )

    assert (
        create_body_component_l(data=gt_tbl_built)
        == "4.66 $\\times$ 10\\textsuperscript{5} \\_ & 4.509 \\\\\n-3.45 $\\times$ 10\\textsuperscript{-9} \\_ & 176.23 \\\\"
    )


def test_create_body_component_l_fmt_percent(gt_tbl_pct: GT):
    gt_tbl_built = gt_tbl_pct.fmt_percent(columns="x", pattern="{x} _")._build_data(context="latex")

    assert (
        create_body_component_l(data=gt_tbl_built)
        == "53.00\\% \\_ & 0.17 \\\\\n6.74\\% \\_ & 0.32 \\\\"
    )


def test_create_body_component_l_fmt_currency(gt_tbl_dec: GT):
    gt_tbl_built = gt_tbl_dec.fmt_currency(columns="x", pattern="{x} _")._build_data(
        context="latex"
    )

    assert (
        create_body_component_l(data=gt_tbl_built)
        == "\\$1.52 \\_ & 4.75 \\\\\n\\$2.23 \\_ & 5.23 \\\\"
    )


def test_create_body_component_l_fmt_bytes(gt_tbl_sci: GT):
    gt_tbl_built = gt_tbl_sci.fmt_bytes(columns="x", pattern="{x} _")._build_data(context="latex")

    assert (
        create_body_component_l(data=gt_tbl_built)
        == "465.6 kB \\_ & 4.509 \\\\\n0 B \\_ & 176.23 \\\\"
    )


def test_create_body_component_l_fmt_date(gt_tbl_dttm: GT):
    gt_tbl_built = gt_tbl_dttm.fmt_date(
        columns="date", date_style="wday_month_day_year", pattern="{x} _"
    )._build_data(context="latex")

    assert (
        create_body_component_l(data=gt_tbl_built)
        == "Saturday, August 12, 2023 \\_ & 09:21:23 & 2023-08-12 09:21:23 \\\\\nTuesday, November 17, 2020 \\_ & 22:45:02 & 2020-11-17 22:45:02 \\\\"
    )


def test_create_body_component_l_fmt_time(gt_tbl_dttm: GT):
    gt_tbl_built = gt_tbl_dttm.fmt_time(
        columns="time", time_style="h_m_s_p", pattern="{x} _"
    )._build_data(context="latex")

    assert (
        create_body_component_l(data=gt_tbl_built)
        == "2023-08-12 & 9:21:23 AM \\_ & 2023-08-12 09:21:23 \\\\\n2020-11-17 & 10:45:02 PM \\_ & 2020-11-17 22:45:02 \\\\"
    )


def test_create_body_component_l_fmt_datetime(gt_tbl_dttm: GT):
    gt_tbl_built = gt_tbl_dttm.fmt_datetime(
        columns="dttm", date_style="wday_month_day_year", time_style="h_m_s_p", pattern="{x} _"
    )._build_data(context="latex")

    assert (
        create_body_component_l(data=gt_tbl_built)
        == "2023-08-12 & 09:21:23 & Saturday, August 12, 2023 9:21:23 AM \\_ \\\\\n2020-11-17 & 22:45:02 & Tuesday, November 17, 2020 10:45:02 PM \\_ \\\\"
    )


def test_create_body_component_l_fmt_roman(gt_tbl_dec: GT):
    gt_tbl_built = gt_tbl_dec.fmt_roman(columns="x", rows=0, pattern="{x} _")._build_data(
        context="latex"
    )

    assert create_body_component_l(data=gt_tbl_built) == "II \\_ & 4.75 \\\\\n2.23 & 5.23 \\\\"


def test_create_wrap_start():
    assert create_wrap_start_l(use_longtable=False, tbl_pos=None) == "\\begin{table}[!t]"
    assert create_wrap_start_l(use_longtable=False, tbl_pos="!b") == "\\begin{table}[!b]"
    assert create_wrap_start_l(use_longtable=True, tbl_pos=None) == "\\begingroup"


@mock.patch.dict(os.environ, {"QUARTO_BIN_PATH": "1"}, clear=True)
def test_create_wrap_start_quarto():
    assert create_wrap_start_l(use_longtable=False, tbl_pos="!t") == "\\begin{table}"
    assert create_wrap_start_l(use_longtable=True, tbl_pos="!t") == "\\begingroup"


def test_create_wrap_end_l():
    assert create_wrap_end_l(use_longtable=False) == "\\end{table}"
    assert create_wrap_end_l(use_longtable=True) == "\\endgroup"


def test_create_table_end_l_longtable():
    assert create_table_end_l(use_longtable=False) == "\\bottomrule\n\\end{tabular*}"
    assert create_table_end_l(use_longtable=True) == "\\bottomrule\n\\end{longtable}"


def test_create_table_start_l_longtable(gt_tbl: GT):
    gt_tbl_no_source_notes = gt_tbl._build_data(context="latex")
    gt_tbl_source_notes = gt_tbl.tab_source_note(source_note="Note")._build_data(context="latex")

    assert (
        create_table_start_l(
            data=gt_tbl_no_source_notes,
            use_longtable=True,
        )
        == "\\begin{longtable}{rr}"
    )

    assert (
        create_table_start_l(
            data=gt_tbl_source_notes,
            use_longtable=True,
        )
        == "\\setlength{\\LTpost}{0mm}\n\\begin{longtable}{rr}"
    )


def test_create_table_start_l_float_tbl_pct(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_width="50%")

    assert (
        create_table_start_l(
            data=gt_tbl_new,
            use_longtable=False,
        )
        == "\\begin{tabular*}{0.5\\linewidth}{@{\\extracolsep{\\fill}}rr}"
    )


def test_create_table_start_l_float_tbl_px(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_width="500px")

    assert (
        create_table_start_l(
            data=gt_tbl_new,
            use_longtable=False,
        )
        == "\\begin{tabular*}{375.0pt}{@{\\extracolsep{\\fill}}rr}"
    )


def test_create_table_start_l_float_tbl_auto(gt_tbl: GT):
    gt_tbl_new = gt_tbl.tab_options(table_width="auto")

    assert (
        create_table_start_l(
            data=gt_tbl_new,
            use_longtable=False,
        )
        == "\\begin{tabular*}{\\linewidth}{@{\\extracolsep{\\fill}}rr}"
    )


def test_snap_render_as_latex_longtable(snapshot):
    gt_tbl = (
        GT(
            gtcars[["mfr", "model", "hp", "trq", "msrp"]].head(5),
        )
        .tab_header(title="The _title_", subtitle="The subtitle")
        .tab_spanner(label="Make _and_ Model", columns=["mfr", "model"])
        .tab_spanner(label="Performance", columns=["hp", "trq"])
        .fmt_currency(columns="msrp")
        .tab_source_note("Note 1")
        .tab_source_note("Note 2")
        .tab_options(table_width="600px", table_font_size="12px")
    )

    latex_str = _render_as_latex(
        data=gt_tbl._build_data(context="latex"), use_longtable=True, tbl_pos=None
    )

    assert snapshot == latex_str


def test_snap_render_as_latex_floating_table(snapshot):
    gt_tbl = (
        GT(
            gtcars[["mfr", "model", "hp", "trq", "msrp"]].head(5),
        )
        .tab_header(title="The _title_", subtitle="The subtitle")
        .tab_spanner(label="Make _and_ Model", columns=["mfr", "model"])
        .tab_spanner(label="Performance", columns=["hp", "trq"])
        .fmt_currency(columns="msrp")
        .tab_source_note("Note 1")
        .tab_source_note("Note 2")
        .tab_options(table_width="600px", table_font_size="12px")
    )

    latex_str = _render_as_latex(
        data=gt_tbl._build_data(context="latex"), use_longtable=False, tbl_pos=None
    )

    assert snapshot == latex_str


def test_render_as_latex_stub_raises():
    gt_tbl = GT(exibble, rowname_col="row")
    with pytest.raises(NotImplementedError) as exc_info:
        _render_as_latex(data=gt_tbl._build_data(context="latex"))

    assert (
        "The table stub (row names and/or row groups) are not yet supported in LaTeX output."
        in exc_info.value.args[0]
    )


def test_render_as_latex_rowgroup_raises():
    gt_tbl = GT(exibble, groupname_col="group")
    with pytest.raises(NotImplementedError) as exc_info:
        _render_as_latex(data=gt_tbl._build_data(context="latex"))

    assert "Row groups are not yet supported in LaTeX output." in exc_info.value.args[0]
