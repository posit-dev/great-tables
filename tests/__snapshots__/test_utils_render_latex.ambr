# serializer version: 1
# name: test_snap_render_as_latex
  '''
  \begingroup
  \setlength\LTleft{\dimexpr(0.5\linewidth - 225pt)}
  \setlength\LTright{\dimexpr(0.5\linewidth - 225pt)}
  \fontsize{9.0pt}{10.8pt}\selectfont
  
  \setlength{\LTpost}{0mm}
  \begin{longtable}{@{\extracolsep{\fill}}llrrr}
  \caption*{
  {\large The \_title\_} \\
  {\small The subtitle}
  } \\
  \toprule
  \multicolumn{2}{c}{Make \_and\_ Model} & \multicolumn{2}{c}{Performance} &  \\ 
  \cmidrule(lr){1-2} \cmidrule(lr){3-4}
  mfr & model & hp & trq & msrp \\ 
  \midrule\addlinespace[2.5pt]
  Ford & GT & 647.0 & 550.0 & \$447,000.00 \\
  Ferrari & 458 Speciale & 597.0 & 398.0 & \$291,744.00 \\
  Ferrari & 458 Spider & 562.0 & 398.0 & \$263,553.00 \\
  Ferrari & 458 Italia & 562.0 & 398.0 & \$233,509.00 \\
  Ferrari & 488 GTB & 661.0 & 561.0 & \$245,400.00 \\
  \bottomrule
  \end{longtable}
  \begin{minipage}{\linewidth}
  Note 1\\
  Note 2\\
  \end{minipage}
  \endgroup
  
  '''
# ---
# name: test_snap_render_as_latex_floating_table
  '''
  \begin{table}[!t]
  \caption*{
  {\large The \_title\_} \\
  {\small The subtitle}
  } 
  
  \fontsize{9.0pt}{10.8pt}\selectfont
  
  \begin{tabular*}{450.0pt}{@{\extracolsep{\fill}}llrrr}
  \toprule
  \multicolumn{2}{c}{Make \_and\_ Model} & \multicolumn{2}{c}{Performance} &  \\ 
  \cmidrule(lr){1-2} \cmidrule(lr){3-4}
  mfr & model & hp & trq & msrp \\ 
  \midrule\addlinespace[2.5pt]
  Ford & GT & 647.0 & 550.0 & \$447,000.00 \\
  Ferrari & 458 Speciale & 597.0 & 398.0 & \$291,744.00 \\
  Ferrari & 458 Spider & 562.0 & 398.0 & \$263,553.00 \\
  Ferrari & 458 Italia & 562.0 & 398.0 & \$233,509.00 \\
  Ferrari & 488 GTB & 661.0 & 561.0 & \$245,400.00 \\
  \bottomrule
  \end{tabular*}
  \begin{minipage}{\linewidth}
  Note 1\\
  Note 2\\
  \end{minipage}
  \end{table}
  
  '''
# ---
# name: test_snap_render_as_latex_longtable
  '''
  \begingroup
  \setlength\LTleft{\dimexpr(0.5\linewidth - 225pt)}
  \setlength\LTright{\dimexpr(0.5\linewidth - 225pt)}
  \fontsize{9.0pt}{10.8pt}\selectfont
  
  \setlength{\LTpost}{0mm}
  \begin{longtable}{@{\extracolsep{\fill}}llrrr}
  \caption*{
  {\large The \_title\_} \\
  {\small The subtitle}
  } \\
  \toprule
  \multicolumn{2}{c}{Make \_and\_ Model} & \multicolumn{2}{c}{Performance} &  \\ 
  \cmidrule(lr){1-2} \cmidrule(lr){3-4}
  mfr & model & hp & trq & msrp \\ 
  \midrule\addlinespace[2.5pt]
  Ford & GT & 647.0 & 550.0 & \$447,000.00 \\
  Ferrari & 458 Speciale & 597.0 & 398.0 & \$291,744.00 \\
  Ferrari & 458 Spider & 562.0 & 398.0 & \$263,553.00 \\
  Ferrari & 458 Italia & 562.0 & 398.0 & \$233,509.00 \\
  Ferrari & 488 GTB & 661.0 & 561.0 & \$245,400.00 \\
  \bottomrule
  \end{longtable}
  \begin{minipage}{\linewidth}
  Note 1\\
  Note 2\\
  \end{minipage}
  \endgroup
  
  '''
# ---
